### PR TITLE
[jest]: add features released in v29.4

### PR DESCRIPTION
**Pull request by mrazauskas**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: see below.

### Summary

Jest v29.4 was released earlier today. This PR is adding typings of new API which are shipped with the new version:

- `jest.isEnvironmentTornDown()` added in https://github.com/facebook/jest/pull/13741
- `jest.isolateModulesAsync()` added in https://github.com/facebook/jest/pull/13680
- ~~`jest.replaceProperty()` and `jest.Replaced` added in https://github.com/facebook/jest/pull/13496~~

### CODE REVIEW

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -430,7 +457,8 @@ declare namespace jest {
     type ConstructorArgumentsOf<T> = T extends new (...args: infer A) => any ? A : never;
     type ConstructorReturnType<T> = T extends new (...args: any) => infer C ? C : any;
 
-    interface MockWithArgs<T extends MockableFunction> extends MockInstance<ReturnType<T>, ArgumentsOf<T>, ConstructorReturnType<T>> {
+    interface MockWithArgs<T extends MockableFunction>
+        extends MockInstance<ReturnType<T>, ArgumentsOf<T>, ConstructorReturnType<T>> {
```
Comment (mrazauskas) [inline types/jest/index.d.ts:461]: Prettier cleanup up some formatting here and below.

**Comment by boginw** [comment]
@mrazauskas can you give a quick explanation for why `types/jest-when/index.d.ts` needs to update Minimum TypeScript Version?

**Comment by mrazauskas** [comment]
> can you give a quick explanation for why `types/jest-when/index.d.ts` needs to update Minimum TypeScript Version?

Thanks for asking. Indeed I am not a fan to change TS version in so many packages. Here is the error pushing me to do so:

Screenshot 2023-01-24 at 16 12 31

I looked for a work around with no luck so far. So decided to make the test pass and perhaps someone smarter than me will come up with a better solution.

**Comment by mrazauskas** [comment]
And now CI gives: `mapbox__aws-sdk-jest` Out of memory: failed

My lucky day ;D Have to make a break.

**Comment by mrazauskas** [comment]
Unfortunately I can’t find a way to implement `jest.replaceProperty()`. Types copied from Jest repo gave this error:

Screenshot 2023-01-24 at 16 12 31

As you can see from the commits, bumping the version as suggested leads to changing many other packages. After I bump version in all the packages, CI gave this:

Screenshot 2023-01-24 at 19 50 36

As far as I know, there is no way to acquire more memory from GitHub. So I must leave `jest.replaceProperty()` for smarter than me.

**Review by sheetalkamat — APPROVED** [review]

**Comment by mrazauskas** [comment]
Ready to merge

**Pull request merged**